crates/shared/src/price_estimation/native/oneinch.rs

```diff
@@ -0,0 +1,237 @@
+use {
+    super::{NativePriceEstimateResult, NativePriceEstimating},
+    crate::price_estimation::{
+        Estimate,
+        PriceEstimateResult,
+        PriceEstimating,
+        PriceEstimationError,
+        Query,
+    },
+    anyhow::{anyhow, Result},
+    futures::{future::BoxFuture, FutureExt},
+    model::order::OrderKind,
+    number::nonzero::U256 as NonZeroU256,
+    primitive_types::{H160, U256},
+    reqwest::Client,
+    std::{
+        collections::HashMap,
+        sync::{Arc, Mutex},
+        time::Duration,
+    },
+};
+
+const BASE_URL: &str = "https://api.1inch.dev/";
+const REFRESH_INTERVAL: Duration = Duration::from_secs(12);
+
+struct OneInch {
+    client: Client,
+    chain_id: u64,
+    // Denominated in wei
+    prices: Arc<Mutex<HashMap<H160, U256>>>,
+    native_price_estimation_amount: NonZeroU256,
+    native_token: H160,
+}
+
+impl OneInch {
+    #[allow(dead_code)]
+    pub fn new(
+        client: Client,
+        chain_id: u64,
+        native_price_estimation_amount: NonZeroU256,
+        native_token: H160,
+    ) -> Self {
+        let instance = Self {
+            client,
+            chain_id,
+            prices: Arc::new(Mutex::new(HashMap::new())),
+            native_price_estimation_amount,
+            native_token,
+        };
+        instance.update_prices_in_background();
+        instance
+    }
+
+    fn update_prices_in_background(&self) {
+        let client = self.client.clone();
+        let chain = self.chain_id;
+        let prices = self.prices.clone();
+        tokio::task::spawn(async move {
+            loop {
+                match update_prices(&client, chain).await {
+                    Ok(new_prices) => {
+                        *prices.lock().unwrap() = new_prices;
+                    }
+                    Err(err) => {
+                        tracing::warn!(?err);
+                    }
+                }
+                tokio::time::sleep(REFRESH_INTERVAL).await;
+            }
+        });
+    }
+}
+
+impl NativePriceEstimating for OneInch {
+    fn estimate_native_price<'a>(
+        &'a self,
+        token: &'a H160,
+    ) -> BoxFuture<'_, NativePriceEstimateResult> {
+        async move {
+            let prices = self.prices.lock().unwrap();
+            let price = prices
+                .get(token)
+                .ok_or_else(|| PriceEstimationError::NoLiquidity)?;
+            Ok(price.to_f64_lossy() / 1e18)
+        }
+        .boxed()
+    }
+}
+
+impl PriceEstimating for OneInch {
+    fn estimate(&self, query: Arc<Query>) -> BoxFuture<'_, PriceEstimateResult> {
```

Comment:
> Currently we do not have a RacingCompetitionPriceEstimator component for NativePriceEstimating (only PriceEstimating) and I'm not sure what the best way is to implement it without repeating this non trivial component.

Since you are already implementing `PriceEstimating` it should be no problem because the `NativePriceEstimator` adapter is currently wrapped around a regular `RacingCompetitionPriceEstimator` using this implementation. So at the moment there is only a single component implementing `NativePriceEstimating`.

Alternatively I think it should be doable to make `RacingCompetitionPriceEstimator` generic over the API (`native` vs `regular`) to make it usable with either estimator type. This could be done by making the inner implementation generic enough to use `impl Fn()` types and have 2 variants of the estimator wrap the core logic (1 for `native`, 1 for `regular`).